### PR TITLE
Issue fixes 1

### PR DESCRIPTION
**Pull request by kfattila**
This PR contains fixes for GitHub issues: 597, 595, 589, 564, 537

### CODE REVIEW

File: src/app/TideIndexApplication.cpp [hunk 1]
```diff
@@ -126,10 +126,11 @@ int TideIndexApplication::main(
       "tide-index.peptides.txt").c_str(), NULL, overwrite);
   }
   
-  ofstream* out_decoy_fasta = GeneratePeptides::canGenerateDecoyProteins() ?
+/*  TODO: Reocer the option to generate decoy protien fasta file.
```
Comment (bmx8177) [inline src/app/TideIndexApplication.cpp:129]: I'm confused about this TODO. I think the typo is confusing me. What do you want to do with the option?

Also - typo in "protien"

File: src/util/Params.cpp [hunk 2]
```diff
@@ -821,6 +821,9 @@ Params::Params() : finalized_(false) {
     "neutral loss peaks have an intensity 1/10 of the primary peak. Neutral losses "
     "are not yet implemented for the res-ev score function.",
     "Available for tide-search.", true);
+  InitIntParam("min-precursor-charge", 1, 1, BILLION,
```
Comment (bmx8177) [inline src/util/Params.cpp:824]: It looks like you added the "min-precursor-charge" parameter but did not remove "spectrum-charge" parameter.

File: src/app/TideSearchApplication.h [hunk 3]
```diff
@@ -320,7 +318,8 @@ class TideSearchApplication : public CruxApplication {
             spectrum_filename(spectrum_filename_), spec_charges(spec_charges_), active_peptide_queue(active_peptide_queue_),
             proteins(proteins_), locations(locations_), precursor_window(precursor_window_), window_type(window_type_),
             spectrum_min_mz(spectrum_min_mz_), spectrum_max_mz(spectrum_max_mz_), min_scan(min_scan_), max_scan(max_scan_),
-            min_peaks(min_peaks_), search_charge(search_charge_), top_matches(top_matches_), highest_mz(highest_mz_),
+            // min_peaks(min_peaks_), search_charge(search_charge_), top_matches(top_matches_), highest_mz(highest_mz_),
```
Comment (bmx8177) [inline src/app/TideSearchApplication.h:321]: I think you can delete this comment.

**Comment by bmx8177** [comment]
It looks like the tide p-value performance test changed a fair bit. Can you remind me what cause this change (I assume a previous merge).

File: src/app/tide/peptide_mods3.cc [hunk 4]
```diff
@@ -61,275 +61,251 @@ class IModsOutputter {
...
-// Alternative class to generate modified peptides. Writes to temporary files
-// before merging them. Bins modified peptides by mass, then writes them to the
-// temporary file associated with that bin; the number of temporary files
-// required is therefore bounded by the variance in peptide masses, rather than
-// the number of possible modifications.
-class ModsOutputterAlt : public IModsOutputter {
+// Original class to generate modified peptides. Writes to temporary files
+// before merging them. As the number of possible modifications increases, the
+// number of required temporary files can grow extremely large.
+class ModsOutputter : public IModsOutputter {
```
Comment (bmx8177) [inline src/app/tide/peptide_mods3.cc:67]: If we are getting rid of the alternate class. Does it make sense to change the class name to the original version? Before Kaipo edited it this line looked like the below.

"class ModsOutputter {"

File: src/app/tide/peptide_mods3.cc [hunk 5]
```diff
@@ -61,275 +61,251 @@ class IModsOutputter {
...
-    } else if (FLAGS_min_mods < 1) {
-      WritePeptide(peptide); // write unmodified peptide
-    }
-    if (maxMods_ == 0) {
-      return;
-    }
-
-    ResultMods resultMods(modTable_, modMaxCounts_, maxMods_, peptide, proteins_);
-    double unmodifiedMass = peptide->mass();
+  uint64_t Total() const { return totalWritten_; }
```
Comment (bmx8177) [inline src/app/tide/peptide_mods3.cc:98]: Is there a reason you remove the following comment?

"// Return the total number of peptides written"

File: src/app/tide/peptide_mods3.cc [hunk 6]
```diff
@@ -61,275 +61,251 @@ class IModsOutputter {
...
 
-    while (resultMods.Next()) {
-      resultMods.ModifyPeptide();
-      WritePeptide(peptide); 
-      if (totalWritten_ % 10000000 == 0) {
-        carp(CARP_INFO, "Wrote %lu modified target peptides to temp files", totalWritten_);
+  void InitCountsMapper() {
+
```
Comment (bmx8177) [inline src/app/tide/peptide_mods3.cc:101]: why is there an empty line here?

File: src/app/tide/peptide_mods3.cc [hunk 7]
```diff
@@ -61,275 +61,251 @@ class IModsOutputter {
...
-        // check that all residues in this combination can be modified
-        bool valid = true;
-        const vector<size_t>& modIndices = modifiedIndices_.Values();
-        for (vector<size_t>::const_iterator i = modIndices.begin(); i != modIndices.end(); i++) {
-          if (modIndexStates_[*i].empty()) {
-            valid = false;
-            break;
+  uint64_t totalWritten_;
+
+  //terminal modifications count as a modification and hence
```
Comment (bmx8177) [inline src/app/tide/peptide_mods3.cc:135]: add space after "//" and before first character in this comment and next comment.

**Comment by bmx8177** [comment]
@kfattila is there a reason why the performance tests do not include combined p-value? Did we have them before?

**Comment by bmx8177** [comment]
Also do you think it is worth it to put a quick description of what the file is supposed to do at the top of src/app/tide/peptide_mods3.cc? It may make it easier for people in the future to understand the script.

File: src/app/tide/peptide_mods3.cc [hunk 8]
```diff
@@ -396,15 +378,17 @@ class ModsOutputterAlt : public IModsOutputter {
...
-  string tempDir_;
   const vector<const pb::Protein*>& proteins_;
-  VariableModTable* modTable_;
-  map<int, int> modMaxCounts_;
-  int maxMods_;
-  HeadedRecordWriter* writer_;
-
-  map< int, pair<string, RecordWriter*> > tempFiles_; // id -> file, writer (ids must be in ascending order of mass)
-  uint64_t totalWritten_;
+  VariableModTable* mod_table_;
```
Comment (bmx8177) [inline src/app/tide/peptide_mods3.cc:382]: why are there variables declared at the beginning and the end of the private class section of ModsOutputter? Doesn't it make more sense for all variables to be together at the beginning of the private section?

File: src/app/TideIndexApplication.cpp (hunk 1 above)
Comment (kfattila) [reply src/app/TideIndexApplication.cpp:129]: I guess, I meant "recover".  The new tide-index does not support an option to produce decoy fasta.

File: src/util/Params.cpp (hunk 2 above)
Comment (kfattila) [reply src/util/Params.cpp:824]: The spectrum-charge option is still used by other programs in CRUX.

File: src/app/TideSearchApplication.h (hunk 3 above)
Comment (kfattila) [reply src/app/TideSearchApplication.h:321]: Right.

File: src/app/tide/peptide_mods3.cc (hunk 6 above)
Comment (kfattila) [reply src/app/tide/peptide_mods3.cc:101]: I removed the empty line.

File: src/app/tide/peptide_mods3.cc (hunk 7 above)
Comment (kfattila) [reply src/app/tide/peptide_mods3.cc:135]: Done.

File: src/app/tide/peptide_mods3.cc (hunk 8 above)
Comment (kfattila) [reply src/app/tide/peptide_mods3.cc:382]: I moved them to the beginning.

**Comment by kfattila** [comment]
"It looks like the tide p-value performance test changed a fair bit. Can you remind me what cause this change (I assume a previous merge)."
I changed the little bit of the null distribution, I think it committed it around January/February. I just updated the performance plot now. The p-value cure actually got better, but somehow percolator's performance decreased on this specific dataset. 

"@kfattila is there a reason why the performance tests do not include combined p-value? Did we have them before?"
I don't know. I did not remove it. I hope. The res-ev test is there.

File: src/app/tide/peptide_mods3.cc (hunk 4 above)
Comment (kfattila) [reply src/app/tide/peptide_mods3.cc:67]: Removed.

File: src/app/tide/peptide_mods3.cc (hunk 5 above)
Comment (kfattila) [reply src/app/tide/peptide_mods3.cc:98]: No idea. I put it back.

**Comment by kfattila** [comment]
"Also do you think it is worth it to put a quick description of what the file is supposed to do at the top of src/app/tide/peptide_mods3.cc? It may make it easier for people in the future to understand the script."

I added a one sentence description. I think, this code is self-explanatory :)

**Comment by kfattila** [comment]
I found another bug, it wrongly printed out some modified decoy peptides. The bug was only in the printing, so it did not affect the performance.

**Pull request merged**